### DISCUSSION TITLE
Slow, slower, slowest

### DISCUSSION THREAD

**Discussion by doobes** [photoprism/photoprism#4638] [Q&A]
So, I stood PhotoPrism up on a HP EliteDesk 705 G3 on Sunday with Proxmox OS. I created a VM with Ubuntu server and Docker to run PP on. I decicated 2 cores and 4Gb of RAM. Maybe that isn't enough....

All my media is on a Synology NAS.  The photos folder on the NAS were mapped to the VM via CIFS. The Docker Compose file I downloaded from you guys was set with all the appropriate information and run appropriately. Everything worked nicely the first time out. Kudos on making that really easy.

I opened the app in my browser and hit "complete rescan".

That was Sunday around 1200. Today, some 45 hours later, it's not half way through scanning the archive. I've got some 80,000 images and accompanying metadata files in the archive. I use Capture One for my pp-ing, and it unfortunately makes a LOT of sidecar files. Probably 3 additional sidecars for every image.

Is this scan speed typical?

Thanks

chris

**Comment by doobes** [comment]
Well, 

I'm now in the "Updating Index" stage.

Guessing by the lack of response that this isn't unusual.

Cheers

**Comment by lunatikk1984** [comment]
> So, I stood PhotoPrism up on a HP EliteDesk 705 G3 on Sunday with Proxmox OS. I created a VM with Ubuntu server and Docker to run PP on. I decicated 2 cores and 4Gb of RAM. Maybe that isn't enough....
> 
> All my media is on a Synology NAS. The photos folder on the NAS were mapped to the VM via CIFS. The Docker Compose file I downloaded from you guys was set with all the appropriate information and run appropriately. Everything worked nicely the first time out. Kudos on making that really easy.
> 
> I opened the app in my browser and hit "complete rescan".
> 
> That was Sunday around 1200. Today, some 45 hours later, it's not half way through scanning the archive. I've got some 80,000 images and accompanying metadata files in the archive. I use Capture One for my pp-ing, and it unfortunately makes a LOT of sidecar files. Probably 3 additional sidecars for every image.
> 
> Is this scan speed typical?
> 
> Thanks
> 
> chris

Hi chris,

I'm also experiencing issues with importing the files. With a Qnap TS-251D (J4025 dual core, 12GB RAM) it takes an average of ~6s to import one image file, according to the log. But what I also noticed when checking the logs is that there are gaps inbetween some image imports. With 70 images imported, in 27 cases there was a gap in between the image imports of 6-9s (with 2 outliers of 13s and 23s).

Now I don't know what's going on here, but my gut feeling tells me that this can't be normal and with 70.000 images to import, this process is going to take more than a week (best case). Is this really normal?

Best regards,
tl

**Reply by doobes** [reply]
Well,

It finished scanning a couple of days ago. It's been updating index... for at least 2 days. CPU is about 75%, but memory utilization is pegged at max.  Disk IO is pretty high as well.

Tomorrow will be a week of this process.

It just seems to me that something is not right.  This shouldn't take a week. Real DAM's don't take anywhere near this long to ingest.

Digikam, for instance had all this done in less than a day.

Would be nice if one of the team members would weigh in here.

**Reply by srett** [reply]
To add a data point, I'm running on an i3-6100 (3.70GHz) on Debian 12 and it takes 4 to 5 seconds per photo, from local SSD storage. So four threads should get you down to ~1 second. If network latency is an issue and you have memory to spare, you might be able to speed it up by using more workers than you have CPU cores, *if* CPU usage is low because of high IO wait for the network share. But that obviously also increases memory requirements. Still, if you're not suffering from  #4557, that should be a viable option to explore.

**Comment by lastzero** [comment]
Hello @doobes,
Hello @lunatikk1984,

> Guessing by the lack of response that this isn't unusual.

On the contrary, indexing should not be that slow, and for most of our users it is actually much faster - except, of course, on unsuitable hardware that does not meet the [system requirements](https://docs.photoprism.app/getting-started/#system-requirements).

As mentioned in our [Troubleshooting Guide](https://docs.photoprism.app/getting-started/troubleshooting/), the [database you are using](https://docs.photoprism.app/getting-started/faq/#should-i-use-sqlite-mariadb-or-mysql), as well as [storage throughput and latency](https://docs.photoprism.app/getting-started/troubleshooting/performance/#storage), have the biggest impact on indexing performance:

- https://docs.photoprism.app/getting-started/troubleshooting/performance/
- https://docs.photoprism.app/getting-started/troubleshooting/docker/#network-storage
- https://docs.photoprism.app/getting-started/faq/#should-i-use-sqlite-mariadb-or-mysql

Unfortunately, you did not provide detailed information about your database, storage/network speed, and volume configuration.

However, running PhotoPrism on a "VM via CIFS" means **many layers of virtualization and non-local storage**, which in itself is a bottleneck that can slow things down.

Also, since you've configured your VM with **only 2 cores**, and PhotoPrism doesn't want to block the entire server with its indexing process, it will only index **one file at a time**. Other software you may have tested for comparison might have been running on your desktop or directly on the server **without any CPU and/or memory limitations**.

Hopefully this feedback (and our documentation) will give you a starting point for optimizing your configuration! ✨

**Comment by doobes** [comment]
Thanks,

I've set PP up with MariaDB, and the remainder is pretty much as laid out in your links.

At this juncture, some 7 days into the process, the thing has pegged my memory allocation, which I set at half the available system memory. 
CPU usage is only list at 50%.

I would kill the process and adjust the memory, but the little man in my head is telling me I would have to start this process over. Letting it finish in it's current configuration might be the most expeditious.

I've got another HP EliteDesk 705 G3 on it's way.  Would PP be better installed under Ubuntu server as the only process?

If so, how might I migrate the database to the new machine?

Appreciate the feedback.

chris

**Reply by lastzero** [reply]
The following documentation describes how to migrate your existing MariaDB database to a new server:
- https://docs.photoprism.app/getting-started/troubleshooting/mariadb/#server-migration

As for performance, our "links" only mention which volumes to configure, but you didn't give any details about the type of storage (hard disk, SSD, CIFS) for each volume and how it is shared with your virtual machine. Also, you said that you have configured 2 cores, which is only the bare minimum you should have - nowhere in our documentation does it say that this is enough for great performance :)

**Comment by doobes** [comment]
Ok,

The machine is:

HP Elitedesk 705 G3 Micro MINI PC AMD Pro 3.0 8GB RAM 256GB SSD. Not the highest powered machine in the world, but just from my futzing with it, significantly more capability than a RaspBerryPI (of which I have 6 running right now).

Storage was mounted, inside the VM as: 

//192.168.10.10/photos /mnt/photos cifs credentials=/home/photoprism/credentials/cifs/creds,vers=3.0,uid=1000,gid=1000 0 0

When I set this up, I used [this](https://docs.photoprism.app/getting-started/) as a guide:

```
System Requirements:
You should host PhotoPrism on a server with at least 2 cores, 3 GB of physical memory, and a 64-bit operating system. Beyond these minimum requirements, the amount of RAM should [match the number of CPU cores]. Indexing large photo and video collections also benefits greatly from [local SSD storage], especially for the database and cache files.
```

It seemed, at the time, that I met these.  If these are the "BARE MINIMUM" requirements, perhaps that should be stated.

The problem at this juncture isn't the CPU.  The process(es) have pegged available memory.

It's also looking like the process has consumed all of the 128Gb of drive that I've allocated to it.

So, re-asking the question,

Based upon the machine I have, and the image archive of ~80k photos on an Synology NAS box attached with a 1Gb network, what would be a good resource allocation? Would this process be more useful "bare-metal" vice under Proxmox or even Docker? I don't have an issue with that. 

Thanks!

**Reply by lastzero** [reply]
Are both the "storage" volume and the "database" directory located on the local SSD, so that only the "originals" folder is located on a remote network drive? How does the VM access the other volumes, also via NFS or CIFS?

- https://docs.photoprism.app/getting-started/docker-compose/#photoprismstorage

Obviously, performance is higher with more storage throughput and IOPS, and more CPU cores, which is why our documentation says "at least". That means "minimum" to me? What other wording would be more explicit to avoid misunderstandings?

- https://docs.photoprism.app/getting-started/#system-requirements

Regarding the size of the "storage" folder, see the following answer in our Getting Started FAQ:

- https://docs.photoprism.app/getting-started/faq/#why-is-my-storage-folder-so-large-what-is-in-it

If you'd like to reduce its size and I/O requirements (thus increasing indexing performance), you can increase JPEG compression (reduce JPEG quality) and reduce the pre-rendered thumbnail size, e.g. from 1920 to 1280:

- https://docs.photoprism.app/user-guide/settings/advanced/#static-and-dynamic-size-limits

**Reply by lastzero** [reply]
Regarding memory usage, see the discussion below - it might be related to a glitch in the VM memory management:
- https://github.com/photoprism/photoprism/discussions/4557

**Reply by lastzero** [reply]
Switching to the built-in (native Go) "imaging" library MAY help if it's a problem with *glibc* memory management and *libvips*:
- https://docs.photoprism.app/getting-started/config-options/#preview-images

**Reply by lastzero** [reply]
Feel free to test our development preview image, which is based on Ubuntu 24.10 instead of 24.04, so the base image might have memory management changes/improvements that fix your problems (also try updating the VM software to see if that fixes it):

- https://docs.photoprism.app/getting-started/updates/#development-preview
- https://docs.photoprism.app/release-notes/#development-preview

**Comment by doobes** [comment]
Whoa!

Lot's to look at.

Lemme digest.  Greatly appreciate the feedback.

Cheers,

**Comment by doobes** [comment]
Well,

It's used up all the storage and I'm dead in the water.  I can't figure out how to expand the storage in Proxmox.

Is there a way to do this without Docker?

May not be this week though.

**Reply by srett** [reply]
> Is there a way to do this without Docker?

Yes, but that's a little more advanced. There are plain tarballs, as well as deb and rpm packages: https://dl.photoprism.app/pkg/linux/
At least the deb package has excessive dependencies though, like mariadb, so if you want to run your DB in another container/VM you might prefer to use the tar release and make sure all the necessary dependencies are installed (ffmpeg etc.) manually. So definitely more work and error prone than just using docker.

Regarding storage, how much did you assign? My photoprism storage folder is 75GB with ~50k photos in my library.

**Reply by doobes** [reply]
I assigned 64GB. It might be that I didn't get the various storage locations set properly. The ease of the downloaded docker compose file lulled me into a sense of complacency and I didn't really dig in as much as I should have.

There a tool in Proxmox that purports to expand it, but it doesn't seem to be affecting the VM.  I know enough about this stuff to be dangerous....

thanks

**Reply by lastzero** [reply]
With modern VM software, it should be possible to create a VM with a large enough disk without immediately using additional disk space. If you do this, you should not have to constantly increase the available space as your library grows (or more pictures are indexed).

**Comment by doobes** [comment]
Well,

I've got PP installed as a single Docker application on a new (to me) EliteDesk 705.

My photo vault is mounted as a CIFS volume.

I'm running a scan right now.

This isn't appreciably faster than when I was running under Proxmox....

It's taking 10-15 seconds per image.  At 70k images this is going to take a little minute.

**Reply by srett** [reply]
How's the process coming along? Did you adjust the worker count before starting the indexing process? 15 seconds does sound slower than it should be, but then again I never tested using a network share.

**Comment by doobes** [comment]
Dunno what a worker count is.

It's "Updating index..." at present. Faster than before, but nothing to get excited about.

CPU usage is pretty low.  I'm guessing this is bandwidth limited.

Memory is about 30% of available.

I think the drive is about 11% consumed. I

**Reply by srett** [reply]
https://docs.photoprism.app/getting-started/config-options/#indexing
You can add it in your compose.yml to override the default, which it seems is half the amount of cores the system has. The value will be limited to the number of logical CPU cores though, so my idea from https://github.com/photoprism/photoprism/discussions/4638#discussioncomment-11429672 cannot be tested. 
Keep an eye on memory usage though, a conservative estimate is that doubling the number of worker threads will double memory usage.

**Comment by doobes** [comment]
Well,

Third time is a charm.

As it turns out I had set Ubuntu server up on a VLM of only 100Gb.

That got filled up, and I was back to where I was before, dead in the water.

So, reinstall Ubuntu server, use the WHOLE SSD and she's wailing away as we speak.

I did do a NFS share vice a CIFS share this time to see if it makes any difference.

It does seem to be somewhat faster. 2-5 seconds per image.

cheers

**Comment by doobes** [comment]
Wow.

Things have slowed down measurably.

CPU usage pretty high.

Storage still only at 15%.

I'm maybe 25% done.

From htop:

![image](https://github.com/user-attachments/assets/c437eb8a-b884-4b17-8956-247c10a5db0d)

**Comment by doobes** [comment]
H'mmm,

Maybe half way done scanning. 42% of my 256gb drive is consumed. Will it be enough?

Is it normal for this to consume so much drive space? I've got ~80k images. 

If so, how in any world would this app be run on a RaspberryPi?

So many questions, so little time.

**Comment by doobes** [comment]
Well it's done.

Finished sometime last night.

Time to look around.
